pkg/threadas/threadas_test.go

```diff
@@ -0,0 +1,192 @@
+		t.Skip("Skipping -- test requires root")
+	}
+
+	targetUids := uidData{Uid: targetUid, Euid: targetUid, Gid: targetGid, Egid: targetGid}
+	fnTargetUidsEqual := targetUids.GenerateTestFunc(t, "matches target user", assert.Equal)
+	fnTargetUidsNotEqual := targetUids.GenerateTestFunc(t, "does not matches target user", assert.NotEqual)
+
+	myUids := getUidData()
+	fnMyUidsEqual := myUids.GenerateTestFunc(t, "matches my user", assert.Equal)
+	fnMyUidsNotEqual := myUids.GenerateTestFunc(t, "does not matches my user", assert.NotEqual)
+
+	// This runs a bunch of tests in parallel. This shows that the
+	// spawn thread, drop permissions, and return to normal
+	// behaves as espected. Be somewhat thoughtful in how parallel
+	// works here -- using t.Parallel will spawn a new thread,
+	// which potentially undermines some of what we're
+	// testing. But we do want the top level to be parallel, so
+	// that we have more thread churn.
+	for i := 1; i < 100; i++ {
+		t.Run("", func(t *testing.T) {
+			t.Parallel()
+
+			t.Run("baseline", func(t *testing.T) {
+				require.NoError(t, fnMyUidsEqual(), "no thread, matches my uid")
+				require.NoError(t, fnTargetUidsNotEqual(), "no thread, does not match target")
+			})
+
+			t.Run("change to target uid", func(t *testing.T) {
+				require.NoError(t, ThreadAs(fnTargetUidsEqual, timeout, uint32(targetUid), uint32(targetGid)), "match target")
+				require.NoError(t, ThreadAs(fnMyUidsNotEqual, timeout, uint32(targetUid), uint32(targetGid)), "not match mine")
+			})
+
+			t.Run("change to my uid", func(t *testing.T) {
+				require.NoError(t, ThreadAs(fnTargetUidsNotEqual, timeout, uint32(0), uint32(0)), "not match target")
+				require.NoError(t, ThreadAs(fnMyUidsEqual, timeout, uint32(0), uint32(0)), "matches mine")
+			})
+
+			t.Run("baseline after setuids", func(t *testing.T) {
+				require.NoError(t, fnMyUidsEqual(), "no thread, matches my uid")
+				require.NoError(t, fnTargetUidsNotEqual(), "no thread, does not match target")
+			})
+		})
+	}
+}
+
+func TestReturnPerms(t *testing.T) {
+	t.Parallel()
+
+	if syscall.Getuid() != 0 {
+		t.Skip("Skipping -- test requires root")
+	}
+
+	targetUids := uidData{Uid: targetUid, Euid: targetUid, Gid: targetGid, Egid: targetGid}
+	myUids := getUidData()
+
+	fn := func() error {
+		// ThreadAs starts with a setuid, so assume that's been called for us
+		targetUids.TestUids(t, "target", assert.Equal)
+
+		// reset permissions
+		require.NoError(t, pthread_setugid_np(KAUTH_UID_NONE, KAUTH_GID_NONE), "reset permissions")
+
+		myUids.TestUids(t, "returned", assert.Equal)
+
+		return nil
+	}
+
+	require.NoError(t, ThreadAs(fn, timeout, uint32(targetUid), uint32(targetGid)))
+
+}
+
+func TestTimeout(t *testing.T) {
+	t.Parallel()
+
+	if syscall.Getuid() != 0 {
+		t.Skip("Skipping -- test requires root")
+	}
+
+	fn := func() error {
+		time.Sleep(10 * time.Second)
+		return nil
+	}
+
+	require.ErrorIs(t, ThreadAs(fn, timeout, uint32(0), uint32(0)), &TimeoutError{})
```

Comment:
this should let you drop the additional Is method on the error
```suggestion
	require.ErrorIs(t, ThreadAs(fn, timeout, uint32(0), uint32(0)), &TimeoutError{t: timeout})
```